### PR TITLE
fix: prevent undefined variables on destroy component

### PR DESCRIPTION
**Pull request by smorcuend**
This PR contains:

- [X] bugfix
- [ ] feature
- [ ] refactor
- [ ] tests
- [ ] documentation
- [ ] others

### Breaking Changes?

- [ ] yes
- [X] no

### Checklist

- [X] commit messages follow the [Conventional Commits](https://conventionalcommits.org/) pattern
  - A feature commit message is prefixed "feat:"
  - A bugfix commit message is prefixed "fix:"
  - A document related commit is prefixed "docs:"
- [ ] docs have been added / updated (for bug fixes / features)

### Describe Your Changes
When you use dynamic component instantiation appears undefined access errors on method ngDestroy

### Does this PR affects any existing issues?

- [ ] yes
- [X] no

### CODE REVIEW

**Comment by sibiraj-sr** [comment]
Thanks for the PR @smorcuend . Changes looks good. but not sure what/when this error occurs, can you give a small repro step?

**Comment by sibiraj-s** [comment]
Thanks for the PR @smorcuend . Changes looks good. but not sure what/when this error occurs, can you give a small repro step?

**Comment by smorcuend** [comment]
Hi @sibiraj-s, 
I will try to explain the use case. I'm using this library in my proyects (https://github.com/dashjoin/json-schema-form) It's just an angular form generator from JSON schemas.
On this library you can create widgets to render specific field managers.
If you try to delve how the widgets are rendering, you can notice the internal method is DCL. https://angular.io/guide/dynamic-component-loader

My problem is just I'm trying to render ngx-editor as a widget of this library and component at some point is not initialize properly. 
You can view an example of how it's used:
https://github.com/dashjoin/json-schema-form/blob/master/src/app/custom.component.ts

File: src/lib/editor.component.ts [hunk 1]
```diff
@@ -119,6 +119,8 @@ export class NgxEditorComponent implements ControlValueAccessor, OnInit, OnChang
       subscription.unsubscribe();
     });
 
-    this.editor.destroy();
+    if (this.editor) {
+      this.editor.destroy();
```
Comment (sibiraj-s) [inline src/lib/editor.component.ts:123]: This line can be removed, this should be handled by wherever the Editor instance is created.

**Comment by sibiraj-s** [comment]
>  at some point is not initialize properly.

are you not getting errors? when not initialised properly? there is a check here 
https://github.com/sibiraj-s/ngx-editor/blob/ab1f001e69c67af653e6ccf5c15da1165d4a1d23/src/lib/editor.component.ts#L69-L71

File: src/lib/editor.component.ts [hunk 2]
```diff
@@ -119,8 +119,6 @@ export class NgxEditorComponent implements ControlValueAccessor, OnInit, OnChang
       subscription.unsubscribe();
     });
 
-    if (this.editor) {
-      this.editor.destroy();
-    }
+    this.editor.destroy();
```
Comment (sibiraj-s) [inline src/lib/editor.component.ts]: I mentioned this

```
this.editor.destroy()
```

This can be removed

**Review by sibiraj-s — APPROVED** [review]

**Comment by sibiraj-s** [comment]
So far I have not added test cases. I have started adding it since this week. one by one wherever possible. If I could reproduce (can you provide a bare minimal repo) I would be able to add one for this. Or If you can test cases are welcome 🙏

**Comment by smorcuend** [comment]
Sure. I can't provide the mentioned code where i got the error but we can provide the error stack trace just FYI. I think the problem is due to the behavior and encapsulation of code into my app. There are many pieces of view that are destroyed and created all time. Sometimes the encapsulated dynamic component doesn't execute ngOnInit but onDestroy is always called.

This is the structure:
MyView -> ElementList -> Element -> Array<lib-json-schema-form> -> Custom Field -> NgxEditor 

By other hand, i have tried to reproduce the same behavior on demo page of library json-schema-form and the error doesn't appear.
Check it out: https://github.com/dashjoin/json-schema-form/pull/98/files



Error stack trace mentioned above:
`
Uncaught TypeError: Cannot read property 'unsubscribe' of undefined
    at BubbleComponent.ngOnDestroy (ngx-editor.js:1994)
    at executeOnDestroys (core.js:7338)
    at cleanUpView (core.js:7251)
    at destroyViewTree (core.js:7077)
    at destroyLView (core.js:7229)
    at ViewContainerRef.remove (core.js:23144)
    at ViewContainerRef.clear (core.js:23052)
    at JsonSchemaFormComponent.loadComponent (dashjoin-json-schema-form.js:2387)
    at JsonSchemaFormComponent.ngOnInit (dashjoin-json-schema-form.js:1835)
    at callHook (core.js:2512)
    at callHooks (core.js:2483)
    at executeInitAndCheckHooks (core.js:2434)
    at refreshView (core.js:9383)
    at refreshEmbeddedViews (core.js:10488)
    at refreshView (core.js:9392)
    at refreshEmbeddedViews (core.js:10488)
    at refreshView (core.js:9392)
    at refreshEmbeddedViews (core.js:10488)
    at refreshView (core.js:9392)
    at refreshEmbeddedViews (core.js:10488)
    at refreshView (core.js:9392)
    at refreshEmbeddedViews (core.js:10488)
    at refreshView (core.js:9392)
    at refreshComponent (core.js:10534)
    at refreshChildComponents (core.js:9165)
    at refreshView (core.js:9418)
    at refreshEmbeddedViews (core.js:10488)
    at refreshView (core.js:9392)
    at refreshEmbeddedViews (core.js:10488)
    at refreshView (core.js:9392)
    at refreshEmbeddedViews (core.js:10488)
    at refreshView (core.js:9392)
    at refreshComponent (core.js:10534)
    at refreshChildComponents (core.js:9165)
    at refreshView (core.js:9418)
    at refreshEmbeddedViews (core.js:10488)
    at refreshView (core.js:9392)
    at refreshComponent (core.js:10534)
    at refreshChildComponents (core.js:9165)
    at refreshView (core.js:9418)
    at refreshEmbeddedViews (core.js:10488)
    at refreshView (core.js:9392)
    at refreshEmbeddedViews (core.js:10488)
    at refreshView (core.js:9392)
    at refreshEmbeddedViews (core.js:10488)
    at refreshView (core.js:9392)
    at refreshEmbeddedViews (core.js:10488)
    at refreshView (core.js:9392)
    at refreshEmbeddedViews (core.js:10488)
    at refreshView (core.js:9392)
ngOnDestroy @ ngx-editor.js:1994
executeOnDestroys @ core.js:7338
cleanUpView @ core.js:7251
destroyViewTree @ core.js:7077
destroyLView @ core.js:7229
remove @ core.js:23144
clear @ core.js:23052
loadComponent @ dashjoin-json-schema-form.js:2387
ngOnInit @ dashjoin-json-schema-form.js:1835
callHook @ core.js:2512
callHooks @ core.js:2483
executeInitAndCheckHooks @ core.js:2434
refreshView @ core.js:9383
refreshEmbeddedViews @ core.js:10488
refreshView @ core.js:9392
refreshEmbeddedViews @ core.js:10488
refreshView @ core.js:9392
refreshEmbeddedViews @ core.js:10488
refreshView @ core.js:9392
refreshEmbeddedViews @ core.js:10488
refreshView @ core.js:9392
refreshEmbeddedViews @ core.js:10488
refreshView @ core.js:9392
refreshComponent @ core.js:10534
refreshChildComponents @ core.js:9165
refreshView @ core.js:9418
refreshEmbeddedViews @ core.js:10488
refreshView @ core.js:9392
refreshEmbeddedViews @ core.js:10488
refreshView @ core.js:9392
refreshEmbeddedViews @ core.js:10488
refreshView @ core.js:9392
refreshComponent @ core.js:10534
refreshChildComponents @ core.js:9165
refreshView @ core.js:9418
refreshEmbeddedViews @ core.js:10488
refreshView @ core.js:9392
refreshComponent @ core.js:10534
refreshChildComponents @ core.js:9165
refreshView @ core.js:9418
refreshEmbeddedViews @ core.js:10488
refreshView @ core.js:9392
refreshEmbeddedViews @ core.js:10488
refreshView @ core.js:9392
refreshEmbeddedViews @ core.js:10488
refreshView @ core.js:9392
refreshEmbeddedViews @ core.js:10488
refreshView @ core.js:9392
refreshEmbeddedViews @ core.js:10488
refreshView @ core.js:9392
refreshComponent @ core.js:10534
refreshChildComponents @ core.js:9165
refreshView @ core.js:9418
refreshComponent @ core.js:10534
refreshChildComponents @ core.js:9165
refreshView @ core.js:9418
refreshEmbeddedViews @ core.js:10488
refreshView @ core.js:9392
refreshComponent @ core.js:10534
refreshChildComponents @ core.js:9165
refreshView @ core.js:9418
refreshEmbeddedViews @ core.js:10488
refreshView @ core.js:9392
refreshComponent @ core.js:10534
refreshChildComponents @ core.js:9165
refreshView @ core.js:9418
refreshEmbeddedViews @ core.js:10488
refreshView @ core.js:9392
refreshComponent @ core.js:10534
refreshChildComponents @ core.js:9165
refreshView @ core.js:9418
renderComponentOrTemplate @ core.js:9482
tickRootContext @ core.js:10708
detectChangesInRootView @ core.js:10733
detectChanges @ core.js:22751
tick @ core.js:29473
(anonymous) @ core.js:29357
invoke @ zone-evergreen.js:364
onInvoke @ core.js:28494
invoke @ zone-evergreen.js:363
run @ zone-evergreen.js:123
run @ core.js:28378
next @ core.js:29356
schedulerFn @ core.js:25848
__tryOrUnsub @ Subscriber.js:183
next @ Subscriber.js:122
_next @ Subscriber.js:72
next @ Subscriber.js:49
next @ Subject.js:39
emit @ core.js:25838
checkStable @ core.js:28431
onLeave @ core.js:28544
onInvokeTask @ core.js:28488
invokeTask @ zone-evergreen.js:398
runTask @ zone-evergreen.js:167
invokeTask @ zone-evergreen.js:480
ZoneTask.invoke @ zone-evergreen.js:469
timer @ zone-evergreen.js:2552
setTimeout (async)
scheduleTask @ zone-evergreen.js:2573
scheduleTask @ zone-evergreen.js:385
scheduleTask @ zone-evergreen.js:210
scheduleMacroTask @ zone-evergreen.js:233
scheduleMacroTaskWithCurrentZone @ zone-evergreen.js:1134
(anonymous) @ zone-evergreen.js:2586
proto.<computed> @ zone-evergreen.js:1449
(anonymous) @ trycatch.js:86
hostReportError @ hostReportError.js:2
__tryOrUnsub @ Subscriber.js:191
next @ Subscriber.js:122
_next @ Subscriber.js:72
next @ Subscriber.js:49
next @ Subject.js:39
emit @ core.js:25838
(anonymous) @ core.js:28520
invoke @ zone-evergreen.js:364
run @ zone-evergreen.js:123
runOutsideAngular @ core.js:28423
onHandleError @ core.js:28520
handleError @ zone-evergreen.js:368
runTask @ zone-evergreen.js:170
invokeTask @ zone-evergreen.js:480
ZoneTask.invoke @ zone-evergreen.js:469
timer @ zone-evergreen.js:2552
setTimeout (async)
scheduleTask @ zone-evergreen.js:2573
scheduleTask @ zone-evergreen.js:385
onScheduleTask @ zone-evergreen.js:272
scheduleTask @ zone-evergreen.js:378
scheduleTask @ zone-evergreen.js:210
scheduleMacroTask @ zone-evergreen.js:233
scheduleMacroTaskWithCurrentZone @ zone-evergreen.js:1134
(anonymous) @ zone-evergreen.js:2586
proto.<computed> @ zone-evergreen.js:1449
(anonymous) @ trycatch.js:86
hostReportError @ hostReportError.js:2
__tryOrUnsub @ Subscriber.js:191
next @ Subscriber.js:122
_next @ Subscriber.js:72
next @ Subscriber.js:49
next @ Subject.js:39
emit @ core.js:25838
checkStable @ core.js:28431
onLeave @ core.js:28544
onInvokeTask @ core.js:28488
invokeTask @ zone-evergreen.js:398
runTask @ zone-evergreen.js:167
invokeTask @ zone-evergreen.js:480
ZoneTask.invoke @ zone-evergreen.js:469
timer @ zone-evergreen.js:2552
setTimeout (async)
scheduleTask @ zone-evergreen.js:2573
scheduleTask @ zone-evergreen.js:385
onScheduleTask @ zone-evergreen.js:272
scheduleTask @ zone-evergreen.js:378
scheduleTask @ zone-evergreen.js:210
scheduleMacroTask @ zone-evergreen.js:233
scheduleMacroTaskWithCurrentZone @ zone-evergreen.js:1134
(anonymous) @ zone-evergreen.js:2586
proto.<computed> @ zone-evergreen.js:1449
(anonymous) @ trycatch.js:86
loadStep @ step-list.component.ts:59
(anonymous) @ step-list.component.ts:53
__tryOrUnsub @ Subscriber.js:183
next @ Subscriber.js:122
_next @ Subscriber.js:72
next @ Subscriber.js:49
_next @ map.js:35
next @ Subscriber.js:49
observe @ Notification.js:20
dispatch @ observeOn.js:25
_execute @ AsyncAction.js:51
execute @ QueueAction.js:20
flush @ AsyncScheduler.js:33
schedule @ QueueAction.js:14
schedule @ Scheduler.js:7
schedule @ AsyncScheduler.js:21
(anonymous) @ ngApollo.js:29
invoke @ zone-evergreen.js:364
onInvoke @ core.js:28494
invoke @ zone-evergreen.js:363
run @ zone-evergreen.js:123
run @ core.js:28378
schedule @ ngApollo.js:29
scheduleMessage @ observeOn.js:30
_next @ observeOn.js:33
next @ Subscriber.js:49
notifySubscription @ Observable.js:135
onNotify @ Observable.js:179
next @ Observable.js:235
(anonymous) @ iteration.js:4
iterateObserversSafely @ iteration.js:4
next @ ObservableQuery.js:21
(anonymous) @ iteration.js:4
iterateObserversSafely @ iteration.js:4
next @ Concast.js:24
notifySubscription @ Observable.js:135
onNotify @ Observable.js:179
next @ Observable.js:235
(anonymous) @ asyncMap.js:13
invoke @ zone-evergreen.js:364
onInvoke @ core.js:28494
invoke @ zone-evergreen.js:363
run @ zone-evergreen.js:123
(anonymous) @ zone-evergreen.js:857
invokeTask @ zone-evergreen.js:399
onInvokeTask @ core.js:28481
invokeTask @ zone-evergreen.js:398
runTask @ zone-evergreen.js:167
drainMicroTaskQueue @ zone-evergreen.js:569
invokeTask @ zone-evergreen.js:484
invokeTask @ zone-evergreen.js:1621
globalZoneAwareCallback @ zone-evergreen.js:1658
load (async)
customScheduleGlobal @ zone-evergreen.js:1773
scheduleTask @ zone-evergreen.js:385
onScheduleTask @ zone-evergreen.js:272
scheduleTask @ zone-evergreen.js:378
scheduleTask @ zone-evergreen.js:210
scheduleEventTask @ zone-evergreen.js:236
(anonymous) @ zone-evergreen.js:1928
(anonymous) @ trycatch.js:137
(anonymous) @ instrument.js:365
(anonymous) @ http.js:1773
_trySubscribe @ Observable.js:42
subscribe @ Observable.js:28
call @ catchError.js:14
subscribe @ Observable.js:23
innerSubscribe @ innerSubscribe.js:67
_innerSub @ switchMap.js:44
_next @ switchMap.js:34
next @ Subscriber.js:49
(anonymous) @ subscribeToArray.js:3
_trySubscribe @ Observable.js:42
subscribe @ Observable.js:28
call @ switchMap.js:15
subscribe @ Observable.js:23
call @ tap.js:16
subscribe @ Observable.js:23
call @ finalize.js:11
subscribe @ Observable.js:23
innerSubscribe @ innerSubscribe.js:67
_innerSub @ mergeMap.js:57
_tryNext @ mergeMap.js:51
_next @ mergeMap.js:34
next @ Subscriber.js:49
(anonymous) @ subscribeToArray.js:3
_trySubscribe @ Observable.js:42
subscribe @ Observable.js:28
call @ mergeMap.js:19
subscribe @ Observable.js:23
call @ filter.js:13
subscribe @ Observable.js:23
(anonymous) @ ngApolloLinkHttp.js:150
Subscription @ Observable.js:197
subscribe @ Observable.js:279
complete @ Concast.js:56
push.VUmj.Concast.start @ Concast.js:79
Concast @ Concast.js:71
push.y0To.QueryManager.getObservableFromLink @ QueryManager.js:473
push.y0To.QueryManager.getResultsFromLink @ QueryManager.js:512
resultsFromLink @ QueryManager.js:627
push.y0To.QueryManager.fetchQueryByPolicy @ QueryManager.js:670
fromVariables @ QueryManager.js:575
push.y0To.QueryManager.fetchQueryObservable @ QueryManager.js:582
(anonymous) @ ObservableQuery.js:312
push.6hsN.Reobserver.reobserve @ Reobserver.js:18
push.AQry.ObservableQuery.reobserve @ ObservableQuery.js:317
push.AQry.ObservableQuery.onSubscribe @ ObservableQuery.js:294
(anonymous) @ ObservableQuery.js:13
Subscription @ Observable.js:197
subscribe @ Observable.js:279
(anonymous) @ subscribeToObservable.js:8
_trySubscribe @ Observable.js:42
subscribe @ Observable.js:28
call @ observeOn.js:14
subscribe @ Observable.js:23
call @ map.js:16
subscribe @ Observable.js:23
fetchData @ step-list.component.ts:48
set selectedProduct @ step-list.component.ts:32
StepListComponent_Template_nb_select_ngModelChange_6_listener @ template.html:6
executeListenerWithErrorHandling @ core.js:15174
wrapListenerIn_markDirtyAndPreventDefault @ core.js:15209
schedulerFn @ core.js:25869
__tryOrUnsub @ Subscriber.js:183
next @ Subscriber.js:122
_next @ Subscriber.js:72
next @ Subscriber.js:49
next @ Subject.js:39
emit @ core.js:25838
viewToModelUpdate @ forms.js:5240
updateControl @ forms.js:2566
(anonymous) @ forms.js:2550
emitSelected @ index.js:24068
handleSingleSelect @ index.js:23940
selectOption @ index.js:23925
handleOptionClick @ index.js:23903
(anonymous) @ index.js:24031
__tryOrUnsub @ Subscriber.js:183
next @ Subscriber.js:122
_next @ Subscriber.js:72
next @ Subscriber.js:49
_next @ Subscriber.js:72
next @ Subscriber.js:49
notifyNext @ switchMap.js:66
_next @ innerSubscribe.js:10
next @ Subscriber.js:49
notifyNext @ mergeMap.js:70
_next @ innerSubscribe.js:10
next @ Subscriber.js:49
next @ Subject.js:39
onClick @ index.js:23003
NbOptionComponent_click_HostBindingHandler @ index.js:23054
executeListenerWithErrorHandling @ core.js:15174
wrapListenerIn_markDirtyAndPreventDefault @ core.js:15209
(anonymous) @ platform-browser.js:582
sentryWrapped @ helpers.js:72
invokeTask @ zone-evergreen.js:399
onInvokeTask @ core.js:28481
invokeTask @ zone-evergreen.js:398
runTask @ zone-evergreen.js:167
invokeTask @ zone-evergreen.js:480
invokeTask @ zone-evergreen.js:1621
globalZoneAwareCallback @ zone-evergreen.js:1647
Show 241 more frames
`

**Comment by sibiraj-s** [comment]
> By other hand, i have tried to reproduce the same behavior on demo page of library json-schema-form and the error doesn't appear.

yeah It shouldn't be an issue at all. As long as the editor component is initialised with editor instance, all should work correctly.

I am fine merging this. changes are simple. But just not satisfied with a fix not knowing the root cause.

I will try to create a demo app with DCL and load the editor dynamically and add some routing logics and try, lets see if this error occurs for me

**Comment by sibiraj-s** [comment]
On the other hand. I suspect a place where bubble component could error out. I will check if that is the error for this.

**Comment by sibiraj-s** [comment]
I have read about Dynamic components. Some blogs mentions the same thing you said `it doesn't execute ngOnInit`. but those blogs are years old. The fix mentioned there is after creating the component

```ts
setImmediate(() => { // can be settimeout too
  ref.changeDetectorRef.detectChanges()
})
```

but when I tried, even without this patch everything is working fine. I used `Angular 11` to test DCL. ngOnInit is called always as expected. `what is the angular version are you using?`

Invoking the lifecycle hook is a must, If that init hook is invoked properly, this error won't occur. and the subscription is must for the editor to function properly. Adding the condition may fix the error, but the editor won't be functioning properly.

**Comment by smorcuend** [comment]
Thanks for research about this problem @sibiraj-s 
Maybe the problem will be on https://github.com/dashjoin/json-schema-form DCL instantiation.
I have tried to insert that code on some other places of json-schema-form but same error appears.
`
setTimeout(() => { // can be settimeout too
  ref.changeDetectorRef.detectChanges()
})
` 
Anyway, i'm using Angular 11.x and my PR mentioned previously works fine with Angular 11 as well.

**Comment by sibiraj-s** [comment]
> Maybe the problem will be on https://github.com/dashjoin/json-schema-form DCL instantiation.

possible.

I am closing this PR now. Let me know if there is something I can do here, I will be glad to look into it. If the issue is with this lib. I would love to get it fixed asap.

Thanks @smorcuend

**Comment by smorcuend** [comment]
Thanks for your time @sibiraj-s 
Finally, i have been testing other project https://github.com/kolkov/angular-editor and none errors are thrown.
I don't know what is the root of the problem.

Anyway, congratulations about your project. That's great code.

**Comment by sibiraj-s** [comment]
Cool 👍 .

Also I tried to create a demo and implement json-schema-form. I was able to reproduce the error.

```
this.jService.registerComponent('wyswig', EditorWidgetComponent)
```

Will see If I need to do something to fix this.

**Comment by sibiraj-s** [comment]
Adding this

```ts
componentRef.changeDetectorRef.detectChanges()
```
at the end of `loadComponent` function here fixes the issue https://github.com/dashjoin/json-schema-form/blob/42865766ed9ba4719def70fb842ed8ba6357459a/projects/dashjoin/json-schema-form/src/lib/json-schema-form.component.ts#L843

**Comment by sibiraj-s** [comment]
I don't know much about the library.  but `loadComponent` is called twice during initialisation one on `OnInit` and other on `OnChanges`. I was able to fix it by

https://github.com/dashjoin/json-schema-form/blob/42865766ed9ba4719def70fb842ed8ba6357459a/projects/dashjoin/json-schema-form/src/lib/json-schema-form.component.ts#L376

changing this line to 

```
    if (changes.switch && !changes.switch.firstChange) {
```

so this changes will be ignored if the value is not change on initial render.

And I have no clue why 
https://github.com/dashjoin/json-schema-form/blob/42865766ed9ba4719def70fb842ed8ba6357459a/projects/dashjoin/json-schema-form/src/lib/json-schema-form.component.ts#L373

ngOnInit is called manually from ngOnChanges. May be they are just treating it as a function.

**Pull request closed without merging**